### ISSUE TITLE
createStateHistory [feature]

### ISSUE THREAD

**Issue by julienvincent** [mjackson/history#386]
I'm sure this has been considered in some form or another, but haven't seen any discussion regarding this (please correct me if I'm wrong and just failed to find any) and so I am opening this to discuss the possibility of adding a state-history creator.

This repo seems to me like the correct place for storing location state (as apposed to redux bindings, or some monkey patching with v4).

Is this a feature on the roadmap? And if not and I were to build this would it be a desirable addition?

**Comment by julienvincent** [comment]
Never mind I did not see there is already a ControlledRouter.

**Comment by julienvincent** [comment]
Actually, after properly looking at ControlledRouter, I still think this is the correct place to store state.

**Comment by mjackson** [comment]
This library already manages `location.state` for you. I don't understand what you're asking.

**Comment by julienvincent** [comment]
Sorry for being a bit vague, what I meant was a way to pull that state out into something like redux of mobx. There is the experimental `ControlledRouter` which enables a controlled component style of state management, but this is a separate router and needs a memory and a browser version of itself.

Would it not make sense to handle this logic in the historyCreator, which can then be used by the different routers?

In other words when creating the history you give it state, and onChange callbacks and the history object then provides the same API to routers? Would this even work with a browser router?

**Comment by mjackson** [comment]
The problem with trying to treat history objects as controlled things is that we don't actually control them. We can try to fake it (like `ControlledRouter` does) but in the end the browser still maintains its own internal history stack with a fairly opaque API; we can only ever guess what's inside it.

For example, I could tell you that the location has changed. And you could decide for some reason that you don't want the location to change, so you tell me instead to revert to the old location. Now, I could just naively update the URL, but then the browser's internal history stack would be in a weird state. Instead, in this situation we need to figure out what `n` is in `history.go(n)` so we can get to the previous position in the stack. Unfortunately, browsers never actually expose the current index in the stack, so we are left guessing.

Currently, this library implements a `revertPop` mechanism that tries hard to satisfy the common use case of "prevent a transition". However, it's not bullet-proof (see #157).

**Comment by julienvincent** [comment]
Ah I see the problem now. Thanks for taking the time to reply 👍
